### ISSUE TITLE
Reduce (or fold) macro

### ISSUE THREAD

**Issue by mnacamura** [bakpakin/Fennel#367]
Fennel provides nice macros `collect` and `icollect`, which enable us to map iterator to a table succinctly.
It might be too much for Fennel's minimalism I guess, but could be nice to have another macro to reduce/fold iterators.
Something like this:
```
(macro reduce1 [iter-tbl accum-expr]
  (let [acc (sym :acc)]
    `(do (var ,acc ,(table.remove iter-tbl 1))
         (each ,iter-tbl
           (match ,accum-expr
             v# (set ,acc v#)))
         ,acc)))
; (reduce1 [1 _ n (ipairs [2 3 4 5])]
;   (* acc n)) ;=> 120

; or

(macro reduce2 [iter-tbl accum-expr]
  (let [acc (table.remove iter-tbl 1)]
    `(do (var ,acc ,(table.remove iter-tbl 1))
         (each ,iter-tbl
           (match ,accum-expr
             v# (set ,acc v#)))
         ,acc)))
; (reduce2 [state "" k v (pairs {:a 1 :b 2 :c 3 :d 4})]
;   (when (= (% v 2) 0)
;     (.. state k))) ;=> "bd"
```

**Comment by technomancy** [comment]
This seems pretty reasonable. The second one which takes a `state` is definitely preferable.

One problem is that it makes it so that returning `nil` from the body will cause the previous accumulator to be used; I don't think that's a good idea because it makes it impossible to make the accumulator become `nil` itself. It should just use the body value for the new accumulator no matter what, I think. If you want `nil` to be treated differently you can do that in the body.

If you want to put a PR together we can discuss it more there.

**Comment by technomancy** [comment]
Fixed in 4f10005
